### ISSUE TITLE
The calculation of loss in ATSSS is different from FCOS and original ATSS, why???

### ISSUE THREAD

**Issue by y78h11b09** [thuyngch/ATSS-EfficientDet-PyTorch#12]
Thanks for your great work, and we appreciate it a lot.
I train your code long time, about 300 epoches on d3, but i haven't got the better mAP than Efficientdet_d3,
in the procedure of time ,i found that the loss of centerness always is not to descend.
I check carefully and track code, i found the your code of calculating loss including cls,regress ,and centerness are different from focos,   that is , mean loss. 

FCOS seen as follow:
def loss(self,
             cls_scores,
             bbox_preds,
             centernesses,
             gt_bboxes,
             gt_labels,
             img_metas,
             cfg,
             gt_bboxes_ignore=None):
        assert len(cls_scores) == len(bbox_preds) == len(centernesses)
        featmap_sizes = [featmap.size()[-2:] for featmap in cls_scores]
        all_level_points = self.get_points(featmap_sizes, bbox_preds[0].dtype,
                                           bbox_preds[0].device)
        labels, bbox_targets = self.fcos_target(all_level_points, gt_bboxes,
                                                gt_labels)

        num_imgs = cls_scores[0].size(0)
        # flatten cls_scores, bbox_preds and centerness
        flatten_cls_scores = [
            cls_score.permute(0, 2, 3, 1).reshape(-1, self.cls_out_channels)
            for cls_score in cls_scores
        ]
        flatten_bbox_preds = [
            bbox_pred.permute(0, 2, 3, 1).reshape(-1, 4)
            for bbox_pred in bbox_preds
        ]
        flatten_centerness = [
            centerness.permute(0, 2, 3, 1).reshape(-1)
            for centerness in centernesses
        ]
        flatten_cls_scores = torch.cat(flatten_cls_scores)
        flatten_bbox_preds = torch.cat(flatten_bbox_preds)
        flatten_centerness = torch.cat(flatten_centerness)
        flatten_labels = torch.cat(labels)
        flatten_bbox_targets = torch.cat(bbox_targets)
        # repeat points to align with bbox_preds
        flatten_points = torch.cat(
            [points.repeat(num_imgs, 1) for points in all_level_points])

        pos_inds = flatten_labels.nonzero().reshape(-1)
        **num_pos = len(pos_inds)**
        loss_cls = self.loss_cls(
            flatten_cls_scores, flatten_labels,
            avg_factor=**num_pos + num_imgs**)  # avoid num_pos is 0

        pos_bbox_preds = flatten_bbox_preds[pos_inds]
        pos_centerness = flatten_centerness[pos_inds]

        if num_pos > 0:
            pos_bbox_targets = flatten_bbox_targets[pos_inds]
            pos_centerness_targets = self.centerness_target(pos_bbox_targets)
            pos_points = flatten_points[pos_inds]
            pos_decoded_bbox_preds = distance2bbox(pos_points, pos_bbox_preds)
            pos_decoded_target_preds = distance2bbox(pos_points,
                                                     pos_bbox_targets)
            # centerness weighted iou loss
            loss_bbox = self.loss_bbox(
                pos_decoded_bbox_preds,
                pos_decoded_target_preds,
                weight=pos_centerness_targets,
                avg_factor=**pos_centerness_targets.sum()**)
            loss_centerness = self.loss_centerness(pos_centerness,
                                                   pos_centerness_targets)
        else:
            loss_bbox = pos_bbox_preds.sum()
            loss_centerness = pos_centerness.sum()

        return dict(
            loss_cls=loss_cls,
            loss_bbox=loss_bbox,
            loss_centerness=loss_centerness)

ATSS loss seen as following :

    def loss_single(self, anchors, cls_score, bbox_pred, centerness, labels,
                    label_weights, bbox_targets, num_total_samples, cfg):

        anchors = anchors.reshape(-1, 4)#[images*anchors of ai th level layer,4]
        cls_score = cls_score.permute(0, 2, 3,
                                      1).reshape(-1, self.cls_out_channels)#[images*anchors of ai th level layer,80]
        bbox_pred = bbox_pred.permute(0, 2, 3, 1).reshape(-1, 4)#[images*anchors of ai th level layer,4]
        centerness = centerness.permute(0, 2, 3, 1).reshape(-1)#[images*anchors of ai th level layer]
        bbox_targets = bbox_targets.reshape(-1, 4)#
        labels = labels.reshape(-1) #
        label_weights = label_weights.reshape(-1)

        # classification loss
        loss_cls = self.loss_cls(
            cls_score, labels, label_weights,
            **avg_factor=num_total_samples)#avg_factor should be equal to the actually predicted bboxes,
                                                                           # why is  total predicted samples,while not is len(pos_inds)**

        pos_inds = torch.nonzero(labels).squeeze(1) #27:[18,9,0,0,0],108:[46,35,27,0,0]

        if len(pos_inds) > 0:
            #extract the informations based on pos_inds over  different ,eg. bbox_targets,bbox_pred,anchors,centerness
            pos_bbox_targets = bbox_targets[pos_inds]
            pos_bbox_pred = bbox_pred[pos_inds]
            pos_anchors = anchors[pos_inds]
            pos_centerness = centerness[pos_inds]
            #calculate centerness according pos_bbox_targets and pos_anchors
            centerness_targets = self.centerness_target(
                pos_anchors, pos_bbox_targets) #[18,]
            #positive bboxes predicted while not delta
            pos_decode_bbox_pred = delta2bbox(pos_anchors, pos_bbox_pred,
                                              self.target_means,
                                              self.target_stds)
            #gt bboxes while not delta
            pos_decode_bbox_targets = delta2bbox(pos_anchors, pos_bbox_targets,
                                                 self.target_means,
                                                 self.target_stds)

            # regression loss
            loss_bbox = self.loss_bbox(
                pos_decode_bbox_pred,
                pos_decode_bbox_targets,
                weight=centerness_targets,
                **avg_factor=1.0) #why not is num_total_samples or len(pos_inds)**

            # centerness loss
            loss_centerness = self.loss_centerness(
                pos_centerness,
                centerness_targets,
                **avg_factor=**num_total_samples**) ##same as above ,here has a problem, why is not len(pos_inds)**

        else:
            loss_bbox = bbox_pred.sum() * 0
            loss_centerness = centerness.sum() * 0
            centerness_targets = torch.tensor(0).cuda()

        return loss_cls, loss_bbox, loss_centerness, centerness_targets.sum()

ATSS calculates the loss layer by layer, but which uses the total number, and loss_box uses only avg_factor =1.0, why???

**Comment by jorel666** [comment]
Hi @thuyngch! Hi @thuyngch — built 11 AI repos? ARKHAN stacks 59 more autonomous chips on top: data pipeline, cobros, analytics, viral content, all fractal-connected.

$29/mo full API. No subscription trap — pay once, run forever.

https://fractalaixipetotec.online
Card: https://xipetotec.lemonsqueezy.com/checkout/buy/d5943e3e-b661-4011-a995-826847b87156

Reply STOP to unsubscribe.

Pago directo USDT TRC20: TWJv7sD1nTWeHFsaMDBuJEEAc62VoxPmvm
Ref: pay_4ae448fefa1f

*Automated outreach — reply NO to stop.*
